src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -296,5 +284,140 @@ internal SafeSslHandle(IntPtr validSslPointer, bool ownsHandle) : base(IntPtr.Ze
+
+            public void SetBio(byte[] buffer, int offset, int length)
+            {
+                _byteArray = buffer;
+                _offset = offset;
+                _bytesAvailable = length;
+            }
+
+            public int Read(Span<byte> output)
+            {
+                var bytesToCopy = Math.Min(output.Length, _bytesAvailable);
+                if (bytesToCopy == 0)
+                    return -1;
+                var span = new Span<byte>(_byteArray, _offset, bytesToCopy);
+                span.CopyTo(output);
+                _offset += bytesToCopy;
+                _bytesAvailable -= bytesToCopy;
+                return bytesToCopy;
+            }
+
+            //Bio is already released by the ssl object
+            private void Dispose(bool isDisposing)
+            {
+                if (_handle.IsAllocated)
+                {
+                    _handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                Dispose(false);
+                GC.SuppressFinalize(this);
+            }
+
+            ~ReadBioBuffer()
+            {
+                Dispose(false);
+            }
+        }
+
+        internal class WriteBioBuffer : IDisposable
+        {
+            private SafeBioHandle _bioHandle;
+            private GCHandle _handle;
+            private byte[] _byteArray;
+            private int _bytesWritten;
+            private bool _isHandshake;
+
+            internal WriteBioBuffer(SafeBioHandle bioHandle)
+            {
+                _bioHandle = bioHandle;
+                _handle = GCHandle.Alloc(this, GCHandleType.Normal);
+                Interop.CustomBio.BioSetGCHandle(_bioHandle, _handle);
+            }
+
+            public void SetBio(byte[] buffer, bool isHandshake)
+            {
+                _byteArray = buffer;
+                _bytesWritten = 0;
+                _isHandshake = isHandshake;
+            }
+
+            public int TakeBytes(out byte[] output)
+            {
+                var bytes = _bytesWritten;
+                output = _byteArray;
+                _bytesWritten = 0;
+                _byteArray = null;
+                return bytes;
+            }
+
+            public int Write(Span<byte> input)
+            {
+                if (_isHandshake)
+                {
+                    if (_byteArray == null)
+                    {
+                        _byteArray = new byte[input.Length];
+                        _bytesWritten = 0;
+                    }
+                    else if (_byteArray.Length - _bytesWritten < input.Length)
+                    {
+                        var oldSpan = new Span<byte>(_byteArray);
+                        _byteArray = new byte[input.Length + _bytesWritten];
+                        oldSpan.CopyTo(_byteArray);
+                    }
+                }
+                var bytesToWrite = Math.Min(input.Length, _byteArray.Length - _bytesWritten);
```

Comment:
Sure, used to the aspnet team telling me to "nit" those spaces, personally I prefer them in, to stop the wall of text effect.